File: lunar-landing-game/lunar-landing-game.py
#!/usr/bin/python3
# -*- coding: UTF-8 -*-
#
# lunar-landing-game.py
# Python program to play Lunar Landing Game
#
import sys, os, math

play_again= 1

while (play_again):
  altitude=2350.0
  velocity=-470.0
  fuel=600
  burn=0
  time_interval=1;
  g=-5.0;

  print("\nLunar Landing Game")
  print("You are in a spacecraft approaching the moon. The onboard computer malfunctioned.")
  print("You have to manually choose the amount of fuel to burn to land with zero velocity.")
  print("Using more fuel, you reduce the velocity more, and vice-versa.\n")

  while (altitude > 0 and fuel > 0):
    print("Altitude=%f\tVelocity=%f\tFuel=%d" % (altitude, velocity, fuel))
    burn= int(input("Amount of fuel to burn [0-75]: "))
    if (burn > 75):
        burn=75
        print("Using the maximum of 75 fuel.")
    if (burn < 0):
        burn=0
        print("Using the minimum of 0 fuel.")
    if (burn > fuel):
        burn=fuel
        print("\nNo more fuel! You can only use %f." % burn)
    fuel-= burn;
    altitude= altitude+velocity*time_interval+(burn+g)*time_interval*time_interval/2.0
    if (altitude < 0):
      print("\nYou crashed at a speed of %f" % velocity)
      break
    velocity=velocity+(burn+g)*time_interval

  if (altitude==0 and velocity==0):
    print("Congratulations: you landed safely, with %d fuel left.\n" % fuel)
  else:
    if (fuel <= 0):
      print("You run out of fuel! You are falling from an altitude of %f.\n" % altitude)
      velocity= -math.sqrt(velocity*velocity-2*g*altitude)
      print("You crashed at a speed of %f.\n" % velocity)

  play_again= (input("Want to play again? [y/n] ")!="n")

File: lunar-landing-game/lunar-landing-game.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main() {
  double altitude=2350;
  double speed=-470;
  int fuel=600;
  int burn=0;
  int time_interval=1;
  double g=-5;

  do {
    printf("Altitude=%f\tSpeed=%f\tFuel=%d\n", altitude, speed, fuel);
    printf("Fuel to burn: ");
    scanf("%d", &burn);
    if (burn > 75) burn=75;
    if (burn < 0) burn=0;
    if (burn > fuel) burn=fuel;
    fuel-= burn;
    altitude= altitude+speed*time_interval+(burn+g)*time_interval*time_interval/2;
    if (altitude < 0) {
      printf("You crashed at speed %f\n", speed);
      break;
    }
    speed=speed+(burn+g)*time_interval;

  } while (altitude > 0 && fuel > 0);

  if (altitude==0 && speed==0)
    printf("Congratulations: you landed safely with %d fuel left\n", fuel);
  else {
    if (fuel <= 0) {
      printf("No fuel\n");
      speed=-sqrt(speed*speed-2*g*altitude);
      printf("You crashed at speed %f\n", speed);
    }
  }
  return 0;
}
